### DISCUSSION TITLE
OMC should search more directories for dlls

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#8596] [Ideas]
OMC is not finding a dll for ExternalMedia on Windows, because it is looking inside `ExternalMedia 4.0.0/Resources/Library/` and inside `ExternalMedia 4.0.0/Resources/Library/win64`, but not inside `ExternalMedia 4.0.0/Resources/Library/mingw64`, which is where the .dll file is actually located.

Reported here on @casella's request, cf. https://github.com/modelica-3rdparty/ExternalMedia/issues/54#issuecomment-1047857214

### Steps to Reproduce
* Download the ExternalMedia v4 snapshot from https://github.com/modelica-3rdparty/ExternalMedia/issues/42#issuecomment-1044251864 and extract it where convenient
* Simulate `ExternalMedia.Test.WaterComparison`
* Observe among others the translation error A
```
[1] 17:02:18 Translation Error
[ExternalMedia.Media.BaseClasses.ExternalTwoPhaseMedium: 192:15-196:29]: External function ‘TwoPhaseMedium_getCriticalTemperature_C_impl‘ could not be found in any of the given shared libraries:
  C:/OpenModelica1.19.0-dev-64bit/bin/ExternalMediaLib.dll
  <snip>/ExternalMedia 4.0.0/Resources/Library/win64/ExternalMediaLib.dll
  <snip>/ExternalMedia 4.0.0/Resources/Library/ExternalMediaLib.dll
  C:/OpenModelica1.19.0-dev-64bit/bin/libExternalMediaLib.dll
  <snip>/ExternalMedia 4.0.0/Resources/Library/win64/libExternalMediaLib.dll
  <snip>/ExternalMedia 4.0.0/Resources/Library/libExternalMediaLib.dll
  C:/OpenModelica1.19.0-dev-64bit/bin/ffi/ExternalMediaLib.dll
  C:/OpenModelica1.19.0-dev-64bit/bin/ffi/libExternalMediaLib.dll
```
and translation warning B
```
  [5] 17:02:18 Translation Warning
[ExternalMedia.Media.BaseClasses.ExternalTwoPhaseMedium: 1019:25-1025:15]: Could not find library ExternalMediaLib in either of:
  C:/OpenModelica1.19.0-dev-64bit/lib//omc/ExternalMediaLib.dll
  /usr/lib//ExternalMediaLib.dll
  /lib//ExternalMediaLib.dll
  /usr/lib//ExternalMediaLib.dll
  /lib//ExternalMediaLib.dll
  C:/OpenModelica1.19.0-dev-64bit/bin//ExternalMediaLib.dll
  C:/OpenModelica1.19.0-dev-64bit/lib//omc/ExternalMediaLib.dll
  C:/OpenModelica1.19.0-dev-64bit/lib//ExternalMediaLib.dll
  <snip>/AppData/Roaming/.openmodelica/binaries/ExternalMedia/ExternalMediaLib.dll
  <snip>/ExternalMedia 4.0.0/Resources/Library/mingw64/ExternalMediaLib.dll
  <snip>/ExternalMedia 4.0.0/Resources/Library/win64/ExternalMediaLib.dll
  <snip>/ExternalMedia 4.0.0/Resources/Library/ExternalMediaLib.dll
  C:/OpenModelica1.19.0-dev-64bit/lib//omc/libExternalMediaLib.a
  /usr/lib//libExternalMediaLib.a
  /lib//libExternalMediaLib.a
  /usr/lib//libExternalMediaLib.a
  /lib//libExternalMediaLib.a
  C:/OpenModelica1.19.0-dev-64bit/bin//libExternalMediaLib.a
  C:/OpenModelica1.19.0-dev-64bit/lib//omc/libExternalMediaLib.a
  C:/OpenModelica1.19.0-dev-64bit/lib//libExternalMediaLib.a
  <snip>/AppData/Roaming/.openmodelica/binaries/ExternalMedia/libExternalMediaLib.a
  <snip>/ExternalMedia 4.0.0/Resources/Library/mingw64/libExternalMediaLib.a
  <snip>/ExternalMedia 4.0.0/Resources/Library/win64/libExternalMediaLib.a
  <snip>/ExternalMedia 4.0.0/Resources/Library/libExternalMediaLib.a
  C:/OpenModelica1.19.0-dev-64bit/lib//omc/libExternalMediaLib.lib
  /usr/lib//libExternalMediaLib.lib
  /lib//libExternalMediaLib.lib
  /usr/lib//libExternalMediaLib.lib
  /lib//libExternalMediaLib.lib
  C:/OpenModelica1.19.0-dev-64bit/bin//libExternalMediaLib.lib
  C:/OpenModelica1.19.0-dev-64bit/lib//omc/libExternalMediaLib.lib
  C:/OpenModelica1.19.0-dev-64bit/lib//libExternalMediaLib.lib
  <snip>/AppData/Roaming/.openmodelica/binaries/ExternalMedia/libExternalMediaLib.lib
  <snip>/ExternalMedia 4.0.0/Resources/Library/mingw64/libExternalMediaLib.lib
  <snip>/ExternalMedia 4.0.0/Resources/Library/win64/libExternalMediaLib.lib
  <snip>/ExternalMedia 4.0.0/Resources/Library/libExternalMediaLib.lib
```
* Note how in the `mingw64` folder, OMC apparently expects slightly different names (no `lib` prefix for the dll, no `.dll` before `.a`). Fair enough, this might actually be wrong in ExternalMedia (I don't know, please confirm, it [seems to work with Dymola](https://github.com/modelica-3rdparty/ExternalMedia/issues/50))
* Rename `mingw64/libExternalMediaLib.dll` to `mingw64/ExternalMediaLib.dll`.

* Resimulate and note how the translation warnings B have disappeared - great, it seems to find the dll now!
* However, the translation errors A are still present, because no dll files exist in `/Library/` (only subfolder per platform), nor in `/Library/win64` (there are `vs*` subfolders), and the `Library/mingw64` folder apparently does _not_ get searched for this error (but the translation _warning_ is gone?!)

### Expected Behavior
OMC should find the dll with the corrected name in `Library/mingw64/ExternalMediaLib.dll` (or even better the original one) and not issue a translation error.

Note: Manually copying the renamed dll from `Library/mingw64`to `Library` as a workaround gets me farther, but then I hit another  OMC error that will probably reported separately after the present issue has been dealt with and some more work on ExternalMedia.

### Version and OS
* OpenModelica Version: OpenModelica v1.19.0-dev-613-gd6e04c0efc (64-bit)
* OS: Windows 10, 64 bit

### Additional Context

The `Library` directory tree of ExternalMedia looks like this:
```
├───linux64
│   ├───clang120
│   │       libExternalMediaLib.4.0.0.dylib
│   │       libExternalMediaLib.dylib
│   │
│   ├───clang130
│   │       libExternalMediaLib.4.0.0.dylib
│   │       libExternalMediaLib.dylib
│   │
│   ├───gcc75
│   │       libExternalMediaLib.so
│   │       libExternalMediaLib.so.4.0.0
│   │
│   └───gcc93
│           libExternalMediaLib.so
│           libExternalMediaLib.so.4.0.0
│
├───mingw32
│       libExternalMediaLib.dll
│       libExternalMediaLib.dll.a
│
├───mingw64
│       libExternalMediaLib.dll
│       libExternalMediaLib.dll.a
│
├───win32
│   ├───vs2015
│   │       ExternalMediaLib.dll
│   │       ExternalMediaLib.lib
│   │
│   ├───vs2017
│   │       ExternalMediaLib.dll
│   │       ExternalMediaLib.lib
│   │
│   ├───vs2019
│   │       ExternalMediaLib.dll
│   │       ExternalMediaLib.lib
│   │
│   └───vs2022
│           ExternalMediaLib.dll
│           ExternalMediaLib.lib
│
└───win64
    ├───vs2015
    │       ExternalMediaLib.dll
    │       ExternalMediaLib.lib
    │
    ├───vs2017
    │       ExternalMediaLib.dll
    │       ExternalMediaLib.lib
    │
    ├───vs2019
    │       ExternalMediaLib.dll
    │       ExternalMediaLib.lib
    │
    └───vs2022
            ExternalMediaLib.dll
            ExternalMediaLib.lib
```

**Comment by mahge** [comment]
This is relatively easy to fix. However, the question is if we should. The [specification (12.9.4)](https://specification.modelica.org/maint/3.5/functions.html#annotations-for-external-libraries-and-include-files) does not list `mingw32/64` as standard directories for Windows. Only `win32/64`. 

If we want to separate `mingw` DLLs from the rest then maybe something like `win64/mingw64/` is probably the way to go. Maybe other people have suggestions. We need to discuss this before it can be _fixed_. 

I am going to convert this to a discussion for now.

**Reply by mahge** [reply]
@casella @adrpo @sjoelund.

**Reply by sjoelund** [reply]
I think we should search the same directories we would for linking, which does look for mingw64 today. We already look for libraries there when linking, so I think we should do the same for function evaluation.

**Reply by mahge** [reply]
Fair enough. I will update it.

**Comment by adrpo** [comment]
We have searched for mingw64 for a while now even if is not specified in the spec. I think we should support it as otherwise you would have clashes for different compilers. 

See also: https://github.com/modelica/ModelicaSpecification/issues/1316

**Comment by casella** [comment]
This is a bit messy. The discussion in [MLS #1316](https://github.com/modelica/ModelicaSpecification/issues/1316) eventually didn't consider mingw (why)? 

If mingw doesn't fit neither Linux nor Windows we should relaunch the proposal to standardize it.

But on top of that, what I never understood is how the _files should actually be named_. Linux uses a `lib` prefix, and `.a` or `.so` extensions for static and shared libraries, so you have `libExternalMedia.a` and `libExternalMedia.so`. Windows puts no prefix and uses .lib for static libraries and .dll for dynamic link libraries. I'm not sure what does mingw use. Are there well-recognized OS-specific standards for that, or should this be stated (or repeated) in the spec?

@adrpo, @mahge, can you shed some light based on your experience?

For sure `libExternalMediaLib.dll.a` looks pretty weird to me:)

**Reply by sjoelund** [reply]
>  Are there well-recognized OS-specific standards for that [...]?

Yes

**Reply by mahge** [reply]
> I'm not sure what does mingw use. 

MinGW uses both as in it will look for both. Since it has to operate in a world with a mix of both conventions, I assume. However, It **prefers** the one with the **lib** prefix if it exists. If not it will look for and accept the non-prefixed one as well.

> For sure libExternalMediaLib.dll.a looks pretty weird to me:)

It does look weird. Those are MinGW's versions of import libs ( counterpart to MSVC's .lib import libs). MinGW decided to differentiate them from other static libs and called them `dll.a`. They are not strictly necessary on MinGW **anymore** as it now allows you to link to the DLL directly. 

However, they are preferred in the search order when looking for a library. So `libExternalMediaLib.dll.a` is in fact the first one MinGW linker will look for if you specify `-lExternalMediaLib` when linking. 

This is the stated search order for `-lExternalMediaLib` 
```sh
libExternalMediaLib.dll.a
ExternalMediaLib.dll.a
libExternalMediaLib.a
ExternalMediaLib.lib
<prefix>ExternalMediaLib.dll (*)
libExternalMediaLib.dll
ExternalMediaLib.dll

# (*) provided by ld option `--dll-search-prefix=<prefix>`.
```

**Reply by bilderbuchi** [reply]
I'm wondering then why we are getting the above translation warning B ("Could not find library ExternalMediaLib") - from the list of search locations, omc does not seem to look for either `libExternalMediaLib.dll` nor `libExternalMediaLib.dll.a` files in any location (both of which exist in the mingw64 subfolder)?

**Reply by mahge** [reply]
This is the MinGW linker's search order. Not OpenModleica's :). It was just for shading lights on how dll.a files are used. 

OpenModelica does it's own search at compile time (Modelica compilation). That is where the warning is coming from. 

https://github.com/OpenModelica/OpenModelica/discussions/8596#discussioncomment-2235790

**Comment by mahge** [comment]
The consensus is to search mingw64 for function evaluation purposes as well since we do search it for linking purposes. I will update it.

**Reply by bilderbuchi** [reply]
Great!
..and mingw32, I presume (on 32bit OM versions), or is that not appropriate?

**Reply by sjoelund** [reply]
I'm not 100% sure it should be mingw64 now that you mention it. The triple is x86_64-w64-mingw32 :)

But 32-bit installer should just be killed off.

**Reply by mahge** [reply]
> ..and mingw32

Of course mingw32 as well. 

> I'm not 100% sure it should be mingw64 now that you mention it. The triple is x86_64-w64-mingw32 :)

We have hardcoded these names in our config headers for Windows. We can change them of course. I guess the triplets are emulated with the hierarchical folders. It should have been `win64/mingw` instead of just `mingw/` but it is too late now :).

**Comment by mahge** [comment]
The second issue (translation warning B) is due to the fact that we look for `libfoo.a` and `libfoo.lib` but only for `foo.dll` and NOT for `libfoo.dll`. I guess we should accept that one as well since some build tools add the lib prefix by default even for DLLs. 

** This search is NOT  for function evaluation purposes. It is a different use case.

**Comment by mahge** [comment]
The first problem (search paths for function evaluation) is fixed with #8602. On Windows we now look in both `win<arch>` and `mingw<arch>` folders. If the library exists in both places, `mingw<arch>` has priority.

**Reply by bilderbuchi** [reply]
Thank you for the quick fix!

**Comment by casella** [comment]
Folks, I'm sorry to step in the discussion, but ExternalMedia is not an _OpenModelica_ library, it's a _Modelica_ library. Hence, it should use (as much as possible) standardized names and library locations, so that it works out of the box with any tool that respects the damn standard on any OS mentioned by the damn standard.

It is not up to the OSMC to standardize the mingw64 directory name (or x86_64-w64-mingw32, or whatever), the same way as it is not to OSMC to decide what is the right syntax for expressions. Furthermore, the current directory names used in ExternalMedia/Resources are not the Gospel, they could/should be changed if necessary, so that the library works out of the box everywhere.

The question is: are there compelling reasons to add another directory mingw64 besides win64 and linux64 in the standardized places to look for libraries of external functions? If so, then we must also start the procedure with MAP-LANG to get it standardized, and I'm happy to take care of that. If not, then we should rather avoid that.

 @mahge, @sjoelund, can you please answer that question?
